Mathlib/GroupTheory/GroupAction/Jordan.lean

```diff
@@ -0,0 +1,834 @@
+  apply not_le.mpr h
+  rw [← ncard_inter_add_ncard_union, H, ncard_univ]
+  exact Nat.le_add_left (Nat.card α) (s ∩ t).ncard
+
+theorem ncard_pigeonhole_compl'
+    (h : s.ncard + t.ncard < Nat.card α) : s ∪ t ≠ ⊤ := by
+  intro h'
+  apply not_le.mpr h
+  rw [← ncard_univ, ← top_eq_univ, ← h']
+  exact ncard_union_le s t
+
+end Set
+
+end PigeonHole
+
+open MulAction SubMulAction Subgroup
+
+open scoped Pointwise
+
+/-- A pretransitivity criterion -/
+theorem IsPretransitive.isPretransitive_ofFixingSubgroup_inter
+    {α : Type*} {G : Type*} [Group G] [MulAction G α] {s : Set α}
+    (hs : IsPretransitive (fixingSubgroup G s) (ofFixingSubgroup G s))
+    {g : G} (ha : s ∪ g • s ≠ ⊤) :
+    IsPretransitive (fixingSubgroup G (s ∩ g • s)) (ofFixingSubgroup G (s ∩ g • s)) := by
+  rw [Ne, Set.top_eq_univ, ← Set.compl_empty_iff, ← Ne, ← Set.nonempty_iff_ne_empty] at ha
+  obtain ⟨a, ha⟩ := ha
+  have ha' : a ∈ (s ∩ g • s)ᶜ := by
+    rw [Set.compl_inter]
+    apply Set.mem_union_left
+    rw [Set.compl_union] at ha
+    apply Set.mem_of_mem_inter_left ha
+  rw [isPretransitive_iff_base (⟨a, ha'⟩ : ofFixingSubgroup G (s ∩ g • s))]
+  rintro ⟨x, hx⟩
+  rw [mem_ofFixingSubgroup_iff, Set.mem_inter_iff, not_and_or] at hx
+  rcases hx with hx | hx
+  · obtain ⟨⟨k, hk⟩, hkax⟩ := hs.exists_smul_eq
+      ⟨a, (by intro ha'; apply ha; apply Set.mem_union_left _ ha')⟩
+      ⟨x, hx⟩
+    use ⟨k, (by
+      rw [mem_fixingSubgroup_iff] at hk ⊢
+      intro y  hy
+      apply hk
+      apply Set.mem_of_mem_inter_left hy)⟩
+    · simp only [← SetLike.coe_eq_coe] at hkax ⊢
+      exact hkax
+  · suffices hg'x : g⁻¹ • x ∈ ofFixingSubgroup G s by
+      suffices hg'a : g⁻¹ • a ∈ ofFixingSubgroup G s by
+        obtain ⟨⟨k, hk⟩, hkax⟩ := hs.exists_smul_eq ⟨g⁻¹ • a, hg'a⟩ ⟨g⁻¹ • x, hg'x⟩
+        use ⟨g * k * g⁻¹, (by
+          rw [mem_fixingSubgroup_iff] at hk ⊢
+          intro y hy
+          simp [← smul_smul, smul_eq_iff_eq_inv_smul g]
+          apply hk
+          rw [← Set.mem_smul_set_iff_inv_smul_mem]
+          exact Set.mem_of_mem_inter_right hy)⟩
+        · simp only [← SetLike.coe_eq_coe] at hkax ⊢
+          simp only [SetLike.val_smul] at hkax ⊢
+          rw [← smul_eq_iff_eq_inv_smul] at hkax
+          change (g * k * g⁻¹) • a = x
+          simp only [← smul_smul]
+          exact hkax
+      rw [mem_ofFixingSubgroup_iff]
+      rw [← Set.mem_smul_set_iff_inv_smul_mem]
+      intro h
+      apply ha
+      apply Set.mem_union_right _ h
+    rw [mem_ofFixingSubgroup_iff]
+    intro h
+    apply hx
+    rw [Set.mem_smul_set_iff_inv_smul_mem]
+    exact h
+
+lemma _root_.SubMulAction.add_encard_ofStabilizer_eq
+    (G : Type*) {α : Type*} [Group G] [MulAction G α] (a : α) :
+    1 + (ofStabilizer G a).carrier.encard = ENat.card α :=  by
+  classical
+  rw [ofStabilizer_carrier]
+  convert Set.encard_add_encard_compl {a}
+  · rw [Set.encard_singleton]
+  · exact (Set.encard_univ α).symm
+
+/- lemma _root_.SubMulAction.add_encard_ofStabilizer_eq'
```

Comment:
Marking this as awaiting-author seems it seems that this PR is not yet waiting for review?

Reply:
Those comments really could be deleted. I did some additional clean-up, sorry for that mess.
Note that there is a Pigeonhole section that should fit into Mathlib.Data.Set.Card, but that file is already > 1100 lines.

Reply:
Perhaps Mathlib.Data.Set.Card should be split, but I do think that these pigeonhole lemmas should be moved elsewhere.

Reply:
The limit seems to be 1500, so those 25 lines won't count.